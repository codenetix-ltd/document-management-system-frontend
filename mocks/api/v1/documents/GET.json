{
  "links": {
    "first": "/api/v1/documents?page=1",
    "last": "/api/v1/documents?page=4",
    "prev": null,
    "next": "/api/v1/documents?page=2"
  },
  "meta": {
    "currentPage": 1,
    "from": 1,
    "lastPage": 8,
    "path": "/api/v1/documents",
    "perPage": 15,
    "to": 15,
    "total": 51
  },
  "data": [
    {
      "name": "Wooden sticks Magnum",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 1,
      "templateId": 1,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Wooden sticks Classic",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 2,
      "templateId": 2,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Stick straight 114 mm",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 3,
      "templateId": 3,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Birchwood sticks 114x10x2 mm",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 4,
      "templateId": 4,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Sticks Butterfly 93 mm RPK",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 5,
      "templateId": 5,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Sticks Butterfly 93 mm RPK",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 6,
      "templateId": 6,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Wooden sticks Mag",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 7,
      "templateId": 1,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Wooden sticks Class",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 8,
      "templateId": 2,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Stick straight",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 9,
      "templateId": 3,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Birchwood sticks",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 10,
      "templateId": 4,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Sticks Butterfly",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 11,
      "templateId": 5,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Wooden sticks",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 12,
      "templateId": 2,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Stick straight new",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 13,
      "templateId": 3,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Birchwood sticks new",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 14,
      "templateId": 4,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    },
    {
      "name": "Sticks Butterfly new",
      "ownerId": 1,
      "substituteDocumentId": 1,
      "attributesValues": [
        {
          "id": 1,
          "value": "Example value"
        }
      ],
      "id": 15,
      "templateId": 5,
      "createdAt": "1503224430",
      "updatedAt": "1503310830"
    }
  ]
}
