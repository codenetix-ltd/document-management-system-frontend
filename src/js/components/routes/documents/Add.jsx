import React, { Component } from 'react';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import Tabs from 'Components/common/Tabs'; // todo: use react-bootstrap's tabs component

import DocumentContent from './partials/DocumentContent';

@autobind
export default class DocumentAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      tabsConfig: []
    };
  }

  componentDidMount() {
    this.setState({
      tabsConfig: [
        {
          title: 'Content',
          active: true,
          content: <DocumentContent />
        }
      ]
    });
  }

  breadcrumbs = [
    { pageName: 'Documents', pageLink: '/documents/list', iconCls: 'fa fa-copy' },
    { pageName: 'Document create', pageLink: '/documents', iconCls: 'fa fa-plus' }
  ];

  render() {
    const { tabsConfig } = this.state;
    return (
      <div>
        <ContentHeader title="Document create" breadcrumbs={this.breadcrumbs} />
        <ContentWrapper noBox>
          <Tabs list={tabsConfig} />
        </ContentWrapper>
      </div>
    );
  }
}
