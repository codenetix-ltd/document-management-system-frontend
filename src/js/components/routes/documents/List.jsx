import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import autobind from 'autobind-decorator';
import moment from 'moment';
import first from 'lodash/first';
import pick from 'lodash/pick';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import Pagination from 'Components/partials/Pagination';

import Prompt from 'Components/common/Prompt';
import AlertMessage from 'Components/common/AlertMessage';
import ErrorMessage from 'Components/common/ErrorMessage';
import SelectableTable from 'Components/common/SelectableTable';

import { DataLink } from 'Components/common/dataControls';

import {
  $$documentArchive,
  $$documentDelete,
  $$documentsFetch
} from 'Store/thunks/documents';

import { $setSelectedDocuments } from 'Store/actions';

import FiltersWrapper from './partials/FiltersWrapper';
import DocumentActions from './partials/DocumentActions';
import ArchiveModalContent from './partials/ArchiveModalContent';

@autobind
export class DocumentsList extends Component {
  static defaultProps = {
    substituteDocument: null
  };

  static propTypes = {
    substituteDocument: PropTypes.any,
    documents: PropTypes.any.isRequired,
    loading: PropTypes.bool.isRequired,
    dispatch: PropTypes.func.isRequired
  };

  componentWillUnmount() {
    this.clearSelection();
  }

  onFetchData(tableProps) {
    let orderBy = 'id';
    let sortedBy = 'desc';
    const { dispatch } = this.props;
    const { page, sorted } = tableProps;
    const sortData = first(sorted);
    if (sortData) {
      orderBy = sortData['id'];
      sortedBy = sortData['desc'] ? 'desc' : 'asc';
    }
    $$documentsFetch(dispatch, {
      page: page + 1,
      orderBy,
      sortedBy
    }, this.clearSelection);
  }

  onSelect(selected) {
    this.props.dispatch($setSelectedDocuments(selected));
  }

  onArchive({ value: { actualVersion } }) {
    const { dispatch } = this.props;
    this.prompt.show({
      width: 555,
      body: <ArchiveModalContent />,
      confirmText: 'Archive',
      onConfirm: close => {
        const { substituteDocument: { id } } = this.props;
        $$documentArchive(dispatch, actualVersion, id, close);
      }
    });
  }

  onDelete({ value: { actualVersion } }) {
    const { dispatch } = this.props;
    this.prompt.show({
      body: `Do you really want to delete document ${actualVersion.name}?`,
      confirmText: 'Delete',
      onConfirm: close => $$documentDelete(dispatch, actualVersion, close)
    });
  }

  clearSelection() {
    this.props.dispatch($setSelectedDocuments([]));
  }

  breadcrumbs = [
    { pageName: 'Documents list', pageLink: '/documents/list', iconCls: 'fa fa-list' }
  ];

  columns = [
    {
      Header: 'Id',
      accessor: 'id',
      maxWidth: 65,
    }, {
      Header: 'Name',
      accessor: 'actualVersion.name',
    }, {
      Header: 'Owner',
      accessor: 'owner.fullName',
    }, {
      Header: 'Template ',
      accessor: 'actualVersion.template.name',
    }, {
      Header: 'Created at',
      accessor: 'createdAt',
      Cell: ({ value }) => moment.unix(value).format('YYYY/MM/DD')
    }, {
      Header: 'Updated at',
      accessor: 'updatedAt',
      Cell: ({ value }) => moment.unix(value).format('YYYY/MM/DD')
    }, {
      Header: 'Actions',
      accessor: '',
      maxWidth: 122,
      sortable: false,
      Cell: rowData => {
        const viewLink = `/documents/view/${rowData.value.id}`;
        const editLink = `/documents/${rowData.value.id}`;
        return (
          <div>
            <Link to={viewLink} className="btn btn-primary btn-xs">
              <i className="fa fa-eye" />
            </Link>
            &nbsp;
            <Link to={editLink} className="btn-success btn btn-xs">
              <i className="fa fa-edit" />
            </Link>
            &nbsp;
            <DataLink cls="btn btn btn-warning btn-xs" data={rowData} onClick={this.onArchive}>
              <i className="fa fa-book" />
            </DataLink>
            &nbsp;
            <DataLink cls="btn-danger btn btn-xs" data={rowData} onClick={this.onDelete}>
              <i className="fa fa-trash" />
            </DataLink>
          </div>
        );
      }
    }
  ];

  render() {
    const { documents, loading } = this.props;
    return (
      <div>
        <ContentHeader title="Documents" breadcrumbs={this.breadcrumbs} />
        <ContentWrapper boxClass="box-success">
          <div className="box-body">
            <FiltersWrapper />
            <hr />
            <DocumentActions prompt={this.prompt} />
            <div className="box-body">
              <div className="row">
                <div className="col-sm-12">
                  <AlertMessage />
                  <ErrorMessage />
                  <SelectableTable
                    manual
                    className="-striped"
                    columns={[...this.columns]}
                    data={documents.list}
                    pages={documents.lastPage}
                    loading={loading}
                    onFetchData={this.onFetchData}
                    onSelect={this.onSelect}
                    defaultPageSize={15}
                    PaginationComponent={Pagination}
                    multiSort={false}
                    resizable={false}
                  />
                </div>
              </div>
            </div>
          </div>
        </ContentWrapper>
        <Prompt ref={p => { this.prompt = p; }} />
      </div>
    );
  }
}

const mapStateToProps = (state) => pick(state, ['documents', 'loading', 'substituteDocument']);

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentsList);
