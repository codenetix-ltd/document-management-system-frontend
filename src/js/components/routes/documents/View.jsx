import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import { Tabs, Tab } from 'react-bootstrap';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';

import AlertMessage from 'Components/common/AlertMessage';
import ErrorMessage from 'Components/common/ErrorMessage';

import {
  $$documentFetch,
  $$documentReset
} from 'Store/thunks/documents';

import ViewContent from './partials/ViewContent';
import DocumentVersions from './partials/DocumentVersions';

@autobind
export class DocumentView extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    document: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      activeKey: 1
    };
  }

  componentDidMount() {
    const { dispatch, match } = this.props;
    const { documentID } = match.params;
    if (documentID) {
      $$documentFetch(dispatch, documentID);
    }
  }

  componentWillUnmount() {
    $$documentReset(this.props.dispatch);
  }

  onTabSelect(activeKey) {
    this.setState({ activeKey });
  }

  breadcrumbs = [
    { pageName: 'Documents', pageLink: '/documents/list', iconCls: 'fa fa-list' },
    { pageName: 'View document', pageLink: '', iconCls: 'fa fa-eye' }
  ];

  render() {
    const { activeKey } = this.state;
    const { match, document, document: { actualVersion: doc } } = this.props;
    return (
      <div>
        <ContentHeader title={`View document ${doc && doc.name}`} breadcrumbs={this.breadcrumbs} />
        <ContentWrapper noBox>
          <AlertMessage />
          <ErrorMessage />
          <Tabs id="tabs" activeKey={activeKey} onSelect={this.onTabSelect} animation={false}>
            <Tab eventKey={1} title="Content">
              <ViewContent document={document} />
            </Tab>
            <Tab eventKey={2} title="Versions">
              <DocumentVersions match={match} />
            </Tab>
          </Tabs>
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ document }) => ({ document });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentView);
