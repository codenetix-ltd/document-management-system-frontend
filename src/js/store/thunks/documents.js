import { API } from 'Config';
import pick from 'lodash/pick';
import pickBy from 'lodash/pickBy';
import identity from 'lodash/identity';

import axios from 'Services/request';

import {
  $loading,
  $document,
  $updateDocument,
  $documentsList,
  $setComparedDocuments
} from 'Store/actions';

import { $$messageSet } from 'Store/thunks/message';
import { $$errorSet } from 'Store/thunks/error';

import initialState from 'Store/reducers/initialState.json';

const fn = () => {};

export const $$documentsFetch = (dispatch, params, callback = fn) => {
  const picked = pick(params, ['page', 'sortField', 'sortDirection']);
  const filters = pickBy(params.filterSet, identity);
  dispatch($loading(true));
  axios.get(API.documents, {
    params: { ...picked, ...filters }
  }).then(({ data }) => {
    dispatch($documentsList({
      list: data.data,
      lastPage: data.meta.lastPage,
      ...picked
    }));
    dispatch($loading(false));
    callback(data);
  }).catch(err => {
    console.trace(err);
    dispatch($loading(false));
    $$errorSet(dispatch, err);
  });
};

export const $$documentFetch = (dispatch, documentID, callback = fn) => {
  dispatch($loading(true));
  axios.get(`${API.documents}/${documentID}`).then(({ data }) => {
    dispatch($document({ ...data }));
    dispatch($loading(false));
    callback(data);
  }).catch(err => {
    console.trace(err);
    dispatch($loading(false));
    $$errorSet(dispatch, err);
  });
};

export const $$documentUpdate = (dispatch, part) => {
  dispatch($updateDocument(part));
};

export const $$documentReset = (dispatch) => {
  dispatch($document(initialState.document));
};

export const $$documentDelete = (dispatch, { id, name }, close = fn) => {
  axios.delete(`${API.documents}/${id}`).then(() => {
    $$documentsFetch(dispatch, 1);
    $$messageSet(dispatch, {
      type: 'success',
      text: `The document ${name} was successfully deleted.`
    });
    close();
  }).catch(err => {
    console.trace(err);
    close();
    $$errorSet(dispatch, err);
  });
};

export const $$documentsMassDelete = (dispatch, ids, close = fn) => {
  const p = Promise.all(ids.map(id => axios.delete(`${API.documents}/${id}`)));
  p.then(() => {
    $$documentsFetch(dispatch, 1);
    $$messageSet(dispatch, {
      type: 'success',
      text: 'Selected documents were successfully deleted.'
    });
    close();
  }).catch(err => {
    console.trace(err);
    close();
    $$errorSet(dispatch, err);
  });
};

export const $$comparedDocumentsFetch = (dispatch, ids, callback = fn) => {
  const p = Promise.all(ids.map(id => axios.get(`${API.documents}/${id}`)));
  p.then(d => {
    dispatch($setComparedDocuments(d.map(({ data }) => data)));
    callback(d);
  }).catch(err => {
    console.trace(err);
  });
};
