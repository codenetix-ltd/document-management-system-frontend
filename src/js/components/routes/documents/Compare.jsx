import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import autobind from 'autobind-decorator';
import { Link } from 'react-router-dom';
import isBool from 'lodash/isBoolean';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';

import Nav from 'react-bootstrap/lib/Nav';
import NavItem from 'react-bootstrap/lib/NavItem';

import { DataLink } from 'Components/common/dataControls';

import { $$comparedDocumentsFetch } from 'Store/thunks/documents';

@autobind
export class DocumentsCompare extends Component {
  static propTypes = {
    location: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    comparedDocuments: PropTypes.array.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      activeKey: 0,
      showAll: true
    };
  }

  componentDidMount() {
    const { search } = this.props.location;
    const params = new URLSearchParams(search);
    const documentIds = params.get('documentIds');
    const ids = documentIds.split(',');
    $$comparedDocumentsFetch(this.props.dispatch, ids);
  }

  onNavClick(activeKey) {
    this.setState({ activeKey });
  }

  onFiltersClick(showAll) {
    this.setState({ showAll });
  }

  getDeviationCells(currentSet, attr, subValue) {
    return currentSet.map(doc => {
      const { value } = doc.attributeValues.find(at => at.id === attr.id);
      return <td key={doc.id}>{value[subValue]}</td>;
    });
  }

  valueDiffers(currentSet, cellID) {
    return currentSet.some(doc => {
      const { value: firstVal } = currentSet[0].attributeValues.find(at => at.id === cellID);
      const { value: currVal } = doc.attributeValues.find(at => at.id === cellID);
      return firstVal !== currVal;
    });
  }

  breadcrumbs = [
    { pageName: 'Documents list', pageLink: '/documents/list', iconCls: 'fa fa-list' },
    { pageName: 'Documents comparison', pageLink: '/compare', iconCls: '' }
  ];

  render() {
    const { activeKey, showAll } = this.state;
    const { comparedDocuments: comparedDocs } = this.props;
    const actualDocs = comparedDocs.map(doc => ({ ...doc.actualVersion, ...doc }));
    const joinedByTemplateID = actualDocs.map(doc => [doc]).reduce((acc, curr) => {
      const ids = acc.map(doc => doc[0].template.id);
      if (ids.includes(curr[0].template.id)) {
        const index = acc.findIndex(doc => doc[0].template.id === curr[0].template.id);
        return acc[index].push(curr[0]) && acc;
      }
      return acc.push(curr) && acc;
    }, []);
    const currentSet = joinedByTemplateID[activeKey];
    const { attributes: attrs } = currentSet ? currentSet[0].template : { attributes: [] };
    return (
      <div>
        <ContentHeader title="Documents comparison" breadcrumbs={this.breadcrumbs} />
        <ContentWrapper noBox>
          <div className="box box-default">
            <div className="box-body">
              <Nav bsStyle="pills" activeKey={activeKey} onSelect={this.onNavClick}>
                {
                  joinedByTemplateID.map((docs, i) => {
                    return (
                      <NavItem key={i} eventKey={i}>{`${docs[0].template.name} (${docs.length})`}</NavItem>
                    );
                  })
                }
              </Nav>
              <p className="compare-actions">
                <span className="compare-show-style">Show only:
                  <span className="compare-show-style-item">
                    <DataLink disabled={showAll} data onClick={this.onFiltersClick}>all characteristics</DataLink>
                  </span>
                  |
                  <span className="compare-show-style-item">
                    <DataLink disabled={!showAll} data={false} onClick={this.onFiltersClick}>differences</DataLink>
                  </span>
                </span>
              </p>
            </div>
          </div>
          <div className="box">
            <div className="box-body">
              <div className="row">
                <div className="col-sm-12">
                  <table className="table table-bordered table-compare">
                    <thead>
                      <tr>
                        <th>&nbsp;</th>
                        {
                          currentSet && currentSet.map(doc => {
                            return (
                              <th key={doc.id} className="table-compare-header-cell">
                                <h2><Link to={`/documents/view/${doc.id}`}>{doc.name}</Link></h2>
                                <span>| {doc.owner.fullName}</span>
                              </th>
                            );
                          })
                        }
                      </tr>
                    </thead>
                    <tbody>
                      {
                        attrs.map(attr => {
                          if (attr.type.name === 'Table') {
                            const firstRow = (
                              <tr key={attr.id}>
                                <td><i className="fa fa-table" /> {attr.name}</td>
                                {currentSet.map(doc => <td key={doc.id} />)}
                              </tr>
                            );
                            const cells = [];
                            attr.data.rows.forEach(row => {
                              row.columns.forEach((col, colIndex) => {
                                cells.push({
                                  id: col.id,
                                  rowTitle: row.name,
                                  colTitle: attr.data.columns[colIndex].name
                                });
                              });
                            });
                            const rows = cells.map(cell => {
                              const valueDiffers = this.valueDiffers(currentSet, cell.id);
                              if (!this.state.showAll && !valueDiffers) return null;
                              return (
                                <tr key={cell.id}>
                                  <td>{`${cell.rowTitle} ${cell.colTitle}`}</td>
                                  {
                                    currentSet.map(doc => {
                                      const { value } = doc.attributeValues.find(at => at.id === cell.id);
                                      return <td key={doc.id}>{value}</td>;
                                    })
                                  }
                                </tr>
                              );
                            });
                            return rows.unshift(firstRow) && rows;
                          } else if (attr.type.name === 'Value with deviations') {
                            const valueCells = this.getDeviationCells(currentSet, attr, 'value');
                            const leftCells = this.getDeviationCells(currentSet, attr, 'left');
                            const rightCells = this.getDeviationCells(currentSet, attr, 'right');
                            return [
                              <tr key={1}><td>{attr.name}</td>{valueCells}</tr>,
                              <tr key={2}><td>Deviation(-)</td>{leftCells}</tr>,
                              <tr key={3}><td>Deviation(+)</td>{rightCells}</tr>
                            ];
                          }
                          return (
                            <tr key={attr.id}>
                              <td>{attr.name}</td>
                              {
                                currentSet.map(doc => {
                                  const { value } = doc.attributeValues.find(at => at.id === attr.id);
                                  if (isBool(value)) {
                                    return <td key={doc.id}>{value ? 'Yes' : 'No'}</td>;
                                  }
                                  return <td key={doc.id}>{value}</td>;
                                })
                              }
                            </tr>
                          );
                        })
                      }
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ comparedDocuments }) => ({ comparedDocuments });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentsCompare);
