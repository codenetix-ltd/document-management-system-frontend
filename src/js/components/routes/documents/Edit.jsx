import axios from 'axios';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { If, Then } from 'qc-react-conditionals/lib';
import { Tabs, Tab } from 'react-bootstrap';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';

import { API } from 'Config';

import { $$documentFetch } from 'Store/thunks/documents';

import DocumentForm from './partials/Form';
import AttributesForm from './partials/AttributesForm';
import DocumentVersions from './partials/DocumentVersions';

@autobind
export class DocumentEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    document: PropTypes.any.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      activeKey: 1,
      newDocumentID: null
    };
    this.breadcrumbs = [
      { pageName: 'Documents', pageLink: '/documents/list', iconCls: 'fa fa-copy' },
      { pageName: 'Edit document', pageLink: '', iconCls: 'fa fa-pencil' }
    ];
  }

  componentDidMount() {
    const { dispatch, match } = this.props;
    const { documentID } = match.params;
    $$documentFetch(dispatch, documentID);
  }

  onTabSelect(activeKey) {
    this.setState({ activeKey });
  }

  onFormSubmit() {
    const { document, match } = this.props;
    const { documentID } = match.params;
    if (documentID) {
      axios.put(API.documents, document).catch(e => {
        console.trace(e);
      });
    } else {
      axios.post(API.documents, document).then(({ data }) => {
        if (!data.id) throw new Error('No id field in response.');
        this.setState({ newDocumentID: data.id });
      }).catch(e => {
        console.trace(e);
      });
    }
  }

  validate() {
    const { document } = this.props;
    return !Object.keys(document).every(key => !!document[key]);
  }

  render() {
    const { activeKey, newDocumentID } = this.state;
    const { document: { actualVersion }, match, match: { params: documentID } } = this.props;
    if (newDocumentID) {
      return <Redirect to={`/documents/${newDocumentID}`} />;
    }
    return (
      <div>
        <ContentHeader title={`Edit document ${actualVersion.name}`} breadcrumbs={this.breadcrumbs} />
        <ContentWrapper noBox>
          <Tabs id="tabs" activeKey={activeKey} onSelect={this.onTabSelect} animation={false}>
            <Tab eventKey={1} title="Content">
              <div className="box-header with-border">
                <h3 className="box-title">General information</h3>
              </div>
              <DocumentForm />
              <If is={actualVersion.template}>
                <Then>
                  <div className="box-header with-border">
                    <h3 className="box-title">Document attributes</h3>
                  </div>
                  <AttributesForm />
                </Then>
              </If>
              <div className="box-footer">
                <button
                  type="button"
                  className="btn btn-success"
                  onClick={this.onFormSubmit}
                  disabled={this.validate()}
                >
                  { documentID ? 'Update' : 'Create' }
                </button>
              </div>
            </Tab>
            <Tab eventKey={2} title="Versions">
              <DocumentVersions match={match} />
            </Tab>
          </Tabs>
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ document }) => ({ document });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentEdit);
