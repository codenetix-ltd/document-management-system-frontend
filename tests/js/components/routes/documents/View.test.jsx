import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { DocumentView } from 'Components/routes/documents/View';

describe('Documents View', () => {
  let wrapper;
  let match;
  let document;

  beforeEach(() => {
    match = {
      params: {
        documentID: 1
      }
    };
    document = {
      actualVersion: {}
    };
    wrapper = shallow(<DocumentView dispatch={() => {}} match={match} document={document} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
