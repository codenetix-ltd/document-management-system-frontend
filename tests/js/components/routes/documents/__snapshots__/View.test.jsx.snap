// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documents View should render correctly 1`] = `
<div>
  <ContentHeader
    breadcrumbs={
      Array [
        Object {
          "iconCls": "fa fa-list",
          "pageLink": "/documents/list",
          "pageName": "Documents",
        },
        Object {
          "iconCls": "fa fa-eye",
          "pageLink": "",
          "pageName": "View document",
        },
      ]
    }
    title="View document undefined"
  />
  <ContentWrapper
    boxClass=""
    noBox={true}
  >
    <withRouter(Connect(AlertMessage)) />
    <withRouter(Connect(ErrorMessage)) />
    <Uncontrolled(Tabs)
      activeKey={1}
      animation={false}
      id="tabs"
      onSelect={[Function]}
    >
      <Tab
        eventKey={1}
        title="Content"
      >
        <DocumentViewContent
          document={
            Object {
              "actualVersion": Object {},
            }
          }
        />
      </Tab>
      <Tab
        eventKey={2}
        title="Versions"
      >
        <Connect(DocumentVersions)
          match={
            Object {
              "params": Object {
                "documentID": 1,
              },
            }
          }
        />
      </Tab>
    </Uncontrolled(Tabs)>
  </ContentWrapper>
</div>
`;
